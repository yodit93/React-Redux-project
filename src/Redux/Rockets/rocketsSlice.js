/* eslint-disable */
/* stylelint-disable */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  rockets: [],
  isLoading: false,
  error: null,
};
const url = 'https://api.spacexdata.com/v4/rockets';
export const fetchRockets = createAsyncThunk('rockets/fetchRockets', async (_, { rejectWithValue }) => {
  try {
    const response = await axios(url);
    return response.data;
  } catch (err) {
    return rejectWithValue('Unable to fetch data');
  }
});
const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    bookRockets: (state, action) => {
      const newRockets = state.rockets.map((rocket) => {
        if (action.payload !== rocket.id) {
          return rocket;
        }
        return {
          ...rocket,
          reserved: true,
        };
      });
      return {
        ...state,
        rockets: newRockets,
      };
    },
    cancelReservation: (state, action) => {
      const newRockets = state.rockets.map((rocket) => {
        if (action.payload !== rocket.id) {
          return rocket;
        }
        return {
          ...rocket,
          reserved: false,
        };
      });
      return {
        ...state,
        rockets: newRockets,
      };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchRockets.fulfilled, (state, action) => {
        const data = action.payload;
        const newRockets = data.map((rocket) => (({
          id: rocket.id,
          name: rocket.name,
          description: rocket.description,
          image: rocket.flickr_images[0],
          reserved: false,
        })));
        return {
          ...state,
          rockets: newRockets,
          isLoading: false,
        };
      })
      .addCase(fetchRockets.rejected, (state, action) => ({
        ...state,
        isLoading: false,
        error: action.payload,
      }));
  },
});

export const { bookRockets, cancelReservation } = rocketsSlice.actions;
export default rocketsSlice.reducer;
