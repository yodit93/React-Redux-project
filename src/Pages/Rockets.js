/* eslint-disable */
/* stylelint-disable */
import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import RocketList from '../Components/RocketLists';
import { fetchRockets } from '../Redux/Rockets/rocketsSlice';

const Rockets = () => {
  const { rockets } = useSelector((store) => store.rockets);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchRockets());
  }, [dispatch]);
  return (
    <ul className="rockets">
      {rockets.map((rocket) => (
        <RocketList key={rocket.id} rocket={rocket} />
      ))}
    </ul>
  );
};

export default Rockets;
