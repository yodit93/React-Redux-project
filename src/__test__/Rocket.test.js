import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Rockets from '../Pages/Rockets';

const mockStore = configureMockStore();
const store = mockStore({
  rockets: {
    rockets: [
      {
        id: '1',
        name: 'Falcon 1',
        description: 'The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009',
        image: 'https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png',
        reserved: false,
      },
    ],
    isLoading: false,
    error: null,
  },
});

describe('Rockets', () => {
  it('should render the Rockets component', () => {
    const tree = renderer.create(<Provider store={store}><Rockets /></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should have a list', () => {
    render(
      <Provider store={store}>
        {' '}
        <Rockets />
      </Provider>,
    );
    const listElement = screen.getByRole('list');
    expect(listElement).toBeInTheDocument();
  });
});
