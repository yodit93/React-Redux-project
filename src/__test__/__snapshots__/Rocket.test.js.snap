// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets should render the Rockets component 1`] = `
Array [
  " ",
  <ul
    className="rockets"
  >
    <li
      className="list"
    >
      <img
        alt=""
        className="img"
        src="https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png"
      />
      <div
        className="info"
      >
        <h3
          className="name"
        >
          Falcon 1
        </h3>
        <p
          className="description"
        >
          The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009
        </p>
        <div
          className="reserve-cont"
        >
          <button
            className="reserve"
            onClick={[Function]}
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </div>
    </li>
  </ul>,
]
`;
